### PR TITLE
[3.4] Two-factor authentication

### PR DESCRIPTION
**Pull request by ViacheslavKlimov**
## Pull Request description

_Issue: https://github.com/thingsboard/thingsboard/issues/6070_

### General overview

Supported 2FA providers are TOTP based (e.g. Google Authenticator, Microsoft Authenticator, FreeOTP, Okta Verify, etc.) and 2FA with an SMS verification code.

In the result, the flow of 2FA set up will look something like this:
1. sysadmin or tenant admin configures desired 2FA providers and some security limitations if needed
2. user goes to his profile settings, chooses desired way of 2FA (among the ones configured by admin), scans a QR code for TOTP 2FA, or enters his phone number for SMS 2FA, then enters a verification code (received from an authenticator app in case of TOTP, or in an SMS message otherwise), and if it is correct, the settings will be successfully saved and applied. 

Regarding the authentication process itself, the flow is obvious:
1. user enters his username and password, as usual
2. if 2FA for account is configured, the user gets prompted to enter verification code received in a specific for provider way (for SMS 2FA he has an option of resending the code if needed)
3. after successfully submitting the code, the user gets authenticated and is now allowed to use the platform

As for adjustment options for available 2FA ways,
- for TOTP based one, an admin can configure the issuer name that will be displayed in an authenticator app:
  photo_2022-03-10_18-57-49
- for SMS 2FA, can customize verification message template (with for now available variables of user email and, of course, a verification code)

As to security options, it is possible to configure the following: two-factor auth API rate limiting (for requesting verification code, as well as for code checking), user lockout after a certain number of unsuccessful verification attempts, verification code expiration (for OTP based providers such as SMS 2FA), and the time allotted for the second authentication factor.

#### Configuration

Configuration is managed through `TwoFactorAuthConfigController`. The controller mostly uses `TwoFactorAuthConfigManager`, which is responsible for managing system- and account-level 2FA settings. 

System 2FA settings are stored as `AdminSettings` in case of sysadmin, and as a tenant attribute in case of tenant admin.
Account configuration for 2FA is stored in the user's `additionalInfo`.

##### API usage
First, [configure](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-b0a2864ff7c851ad8230c8797e3eee8555edfb97dd93b473396a74b11f1142e7R179) `TwoFactorAuthSettings` (on sysadmin or tenant admin level) with the POST to `/api/2fa/settings`. See Swagger documentation for [`TwoFactorAuthSettings`](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-ab29af54666d09ff6e9876ed967e0752b4181a79c573a45c9980126c4fb8a107) and providers configs: [`TotpTwoFactorAuthProviderConfig`](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-8c999549044d52f398613f5f7e3dda15c47eb68c7ac0ea7593ddc2d64c0a12bb) and [`SmsTwoFactorAuthProviderConfig`](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-45d88e8d6c43d67ebf65f6cedb98d1bf54fb1206e825f1fb85d7182e83691d52).
Then, to configure 2FA for account:
1. [Generate](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-b0a2864ff7c851ad8230c8797e3eee8555edfb97dd93b473396a74b11f1142e7R82) account config template with a `POST /api/2fa/account/config/generate` request. This API method is useful for [TOTP 2FA](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-d04ee50b284fe086526729d1c146d557911224886759329d77add473b09e661d): in this case the response contains `authUrl` which is used to generate a QR code on UI to be scanned by an authenticator app.
2. [Submit](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-b0a2864ff7c851ad8230c8797e3eee8555edfb97dd93b473396a74b11f1142e7R124) the adjusted config with POST to `/api/2fa/account/config/submit`. For now, this method is only useful for SMS 2FA. The method should be called after a user enters his phone number: it will send a verification code for a given account config.
3. [Verify and save](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-b0a2864ff7c851ad8230c8797e3eee8555edfb97dd93b473396a74b11f1142e7R140) the config with POST request to `api/2fa/account/config` with a verification code as a request param. 
The method will return a Bad Request status if the provided code is incorrect. 

See Swagger documentation for [`TotpTwoFactorAuthAccountConfig`](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-d04ee50b284fe086526729d1c146d557911224886759329d77add473b09e661d) and [`SmsTwoFactorAuthAccountConfig`](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-33343449180e29a44b0755e4c57507e7e7b2ada79dfe6beb8ce6f640b4e4860b).

#### Authentication

The corresponding controller is `TwoFactorAuthController`.
The controller relies on `TwoFactorAuthService`, that collects beans of `TwoFactorAuthProvider` implementations and routes requests to them depending on the needed provider type. 
`TwoFactorAuthProvider` interface implementations provide functionality of verification code checking and generation of new account configs, and can implement a method of code sending. In the future, to add support for a new 2FA provider, it will be enough just to create an implementation of the interface and mark it as a Component, and `TwoFactorAuthService` will then pick it up.

Besides the new controller, `RestAuthenticationProvider` along with `RestAwareAuthenticationSuccessHandler` were adapted to work with a 2FA: the success handler now issues a pre-verification token instead of a usual token pair if a user has 2FA configured. This token can then only be used for calling `TwoFactorAuthController`'s API methods.

##### API usage
1. Make a call to the regular authentication API with username and password (`POST /api/auth/login`), and get a pre-verification access token in the response: `{ "token": "...", "refreshToken": null, "scope": "PRE_VERIFICATION_TOKEN" }`. As you see, now the scope is being additionally returned so that you now that a user uses 2FA and the token is for verification purposes. Further requests to the auth API must be made with this token. This token is relatively short-lived: default lifetime is 30 minutes.
Then, you may make a `GET /api/auth/2fa/provider/type` to [get the type](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-f696e0fb62e2b53278f17e906093750a764957319a83a1e45027ad70b5e2a79eR96) of 2FA provider that is being used by the user.
2. [Request](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-f696e0fb62e2b53278f17e906093750a764957319a83a1e45027ad70b5e2a79eR59) a verification code with a POST request to `/api/auth/2fa/verification/send`. This API call is only useful in case of SMS 2FA, and will do nothing for other provider types.
You may call this method multiple times if needed, e.g. if a user wants to resend an SMS verification message. 
Also note, that this method, as well as the one for code checking, may be rate limited if the admin configured so, and so Too Many Requests status code can be returned.
3. [Check](https://github.com/thingsboard/thingsboard/pull/6235/files#diff-f696e0fb62e2b53278f17e906093750a764957319a83a1e45027ad70b5e2a79eR73) the code with the POST to `/api/auth/2fa/verification/check` with the code as `verificationCode` request param, and get a regular token pair if it is correct. Otherwise, Bad Request status will be returned with the error message of smth like "Verification code is incorrect". 
Also, if the admin configured user to be blocked after certain number of unsuccessful verification attempts and the user reaches the limit, then an Unathorized status will be returned with the error message saying that the user is blocked.

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.
  
## Front-End feature checklist

- [x] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [x] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [x] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: application/pom.xml [hunk 1]
```diff
@@ -337,6 +337,22 @@
             <artifactId>Java-WebSocket</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.jboss.aerogear</groupId>
+            <artifactId>aerogear-otp-java</artifactId>
+        </dependency>
+        <!-- TMP -->
```
Comment (ashvayka) [inline application/pom.xml:344]: What does TMP mean?

File: application/src/main/java/org/thingsboard/server/controller/BaseController.java [hunk 2]
```diff
@@ -279,11 +282,37 @@
     @Getter
     protected boolean edgesEnabled;
 
+    @ExceptionHandler(Exception.class)
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/BaseController.java:285]: What is the motivation for this?

File: common/message/src/main/java/org/thingsboard/server/common/msg/tools/TbRateLimits.java [hunk 3]
```diff
@@ -29,21 +30,24 @@
...
+    }
+
+    public TbRateLimits(String limitsConfiguration, boolean refillIntervally) {
         LocalBucketBuilder builder = Bucket4j.builder();
         boolean initialized = false;
         for (String limitSrc : limitsConfiguration.split(",")) {
             long capacity = Long.parseLong(limitSrc.split(":")[0]);
             long duration = Long.parseLong(limitSrc.split(":")[1]);
-            builder.addLimit(Bandwidth.simple(capacity, Duration.ofSeconds(duration)));
+            Refill refill = refillIntervally ? Refill.intervally(capacity, Duration.ofSeconds(duration)) : Refill.greedy(capacity, Duration.ofSeconds(duration));
```
Comment (ashvayka) [inline common/message/src/main/java/org/thingsboard/server/common/msg/tools/TbRateLimits.java:42]: Why is this changed? What is the reason and did you add any unit tests?

File: common/dao-api/src/main/java/org/thingsboard/server/dao/user/UserService.java [hunk 4]
```diff
@@ -56,16 +56,19 @@
     PageData<User> findUsersByTenantId(TenantId tenantId, PageLink pageLink);
 
     PageData<User> findTenantAdmins(TenantId tenantId, PageLink pageLink);
-	
+
```
Comment (ashvayka) [inline common/dao-api/src/main/java/org/thingsboard/server/dao/user/UserService.java:59]: Please fix the indentation in this file.

File: application/pom.xml (hunk 1 above)
Comment (ViacheslavKlimov) [reply application/pom.xml:344]: These are temporary libraries for generating a QR code for testing purposes. Will be removed when UI is ready.

File: application/src/main/java/org/thingsboard/server/controller/BaseController.java (hunk 2 above)
Comment (ViacheslavKlimov) [reply application/src/main/java/org/thingsboard/server/controller/BaseController.java:285]: To avoid wrapping each controller's API method in a try-catch block that maps Exception to ThingsboardException.

**Comment by vvlladd28** [comment]
Add new login step page:
image
Add page security for activate 2FA in user:
image
Add new page for admin setting in activate 2FA:
image
